dig/graph.go

```diff
@@ -183,5 +183,38 @@ func (g *graph) registerConstructor(c interface{}) error {
 	}
 
 	g.nodes[objType] = &n
+
+	// object needs to be part of the graph to properly detect cycles
+	if cycleErr := g.detectCycles(&n); cycleErr != nil {
+		// if the cycle was detected delete from the graph
+		delete(g.nodes, objType)
```

Comment:
Not saying that its not the right thing, but whats the outcome of returning an error? This error will be returned, and you have a `nil pointer dereference` at the location where the resolved object is accessed. npd's are pretty vague to debug and doesn't help in understanding the cause. I think Register should be only accessed on initialization, may be not via exported method.